### PR TITLE
Add theoretical security checks to every script available on the lock screen

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:
None

### Summary of the issue:

NVDA may cache or directly access objects below the lock screen while Windows is locked.
As a result, without security checks, secure information may be leaked while Windows is locked.
"Secure objects" refer to `NVDAObjects` which may contain secure information: i.e. objects below the lock screen while Windows is locked.

In 2022.2.4, additional security checks were added to `api.setReviewCursor` to prevent secure objects from being set as the review cursor.
The results of `api.setReviewCursor` are not acknowledged, meaning theoretically a cached secure object may be announced.

Similar issues exist for other scripts available on the lock screen.
A thorough review of every script available on the lock screen is required.

There are no known exploits related to the theoretical issues that this PR attempts to solve.

### Description of user facing changes
None

### Description of development approach
A thorough review of every script available on the lock screen was performed.

Additional security checks were added to ensure that no secure objects or text from secure objects are cached or announced when activating a script available on the lock screen.

### Testing strategy:

Test on alpha for an extensive period, as these changes are minor but widespread and hard to test manually.

### Known issues with pull request:

None

### Change log entries:
None required

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by seanbudd** [comment]
cc @CyrilleB79

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Overall this looks good.

I didn't comment all changes, but it would be good to have an accurate log mesage when the scripts fail due to security. Additionally, for the user, it may be more friendly to supply a specific UI message. 

Additionally, the consistency of the security explainer comments is good, perhaps link them all back to the set of safescripts.

File: source/NVDAObjects/__init__.py [hunk 1]
```diff
@@ -1160,6 +1162,12 @@ def event_mouseMove(self,x,y):
 			info=NVDAObjectTextInfo(self,textInfos.POSITION_FIRST)
 		except LookupError:
 			return
+
+		# This event may fire on the lock screen, as such
+		# ensure the target TextInfo does not contain secure information.
+		if objectBelowLockScreenAndWindowsIsLocked(info.obj):
+			return
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/NVDAObjects/__init__.py:1169]: Logging on failure should be covered in the call to `objectBelowLockScreenAndWindowsIsLocked`.

`ui.message`s will be added where relevant.

File: source/globalCommands.py [hunk 2]
```diff
@@ -1050,9 +1078,15 @@ def script_toggleSimpleReviewMode(self,gesture):
...
-	def script_navigatorObject_current(self,gesture):
+	def script_navigatorObject_current(self, gesture: "InputGesture"):
 		curObject=api.getNavigatorObject()
-		if not isinstance(curObject,NVDAObject):
+		if (
+			not isinstance(curObject, NVDAObject)
+			# This script is available on the lock screen, as such
+			# ensure the navigatorObject does not contain secure information
+			# before announcing this object
+			or objectBelowLockScreenAndWindowsIsLocked(curObject)
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/globalCommands.py:1088]: Logging on failure should be covered in the call to `objectBelowLockScreenAndWindowsIsLocked`.

`ui.message`s will be added where relevant.

File: source/globalCommands.py [hunk 3]
```diff
@@ -949,17 +956,35 @@ def script_cycleSpeechSymbolLevel(self,gesture):
...
-			x=p.x
-			y=p.y
+			reviewPositionStartPoint = None
+
+		if (
+			reviewPositionStartPoint
+			# This script is available on the lock screen, as such
+			# ensure the review position does not contain secure information
+			# before navigating to this object
+			and objectBelowLockScreenAndWindowsIsLocked(reviewPosition.obj)
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/globalCommands.py:971]: Logging on failure should be covered in the call to `objectBelowLockScreenAndWindowsIsLocked`.

`ui.message`s will be added where relevant.

[the remark this answers is not in the document]
Comment (seanbudd) [reply source/globalCommands.py:971]: yes, fixed

**Review by feerrenrut — CHANGES_REQUESTED** [review]
I think it's best to be explicit when that the script is expected to return after checking whether the info can be exposed. Trying to prevent future regressions by making the intent explicit.
I didn't mark all the cases where `return` in missing. It would be safest to ensure every `ui.reviewMessage(gui.blockAction.Context.WINDOWS_LOCKED.translatedMessage)` should be followed by a `return`, and if not comment on why a different pattern is followed.

File: source/globalCommands.py [hunk 4]
```diff
@@ -2111,7 +2364,13 @@ def _getStatusBarText(setReviewCursor: bool = False) -> Optional[str]:
 		"""
 		obj = api.getStatusBar()
 		found = False
-		if obj:
+		if (
+			obj
+			# This script is available on the lock screen via getSafeScripts, as such
+			# ensure the status bar does not contain secure information
+			# before announcing this object
+			and not objectBelowLockScreenAndWindowsIsLocked(obj)
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply source/globalCommands.py:2406]: I have tested reporting the title on the lock screen. "Lock Screen Window" and "Magnifier" were reported correctly from the lock screen and sign-in screen. I'm not sure how gracefully this fails.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**